Add io.github.go_musicfox.go-musicfox

lihaohong6 opened the pull request:
<!-- ⚠️⚠️ Submission pull request MUST be made against the `new-pr` **base branch** ⚠️⚠️  -->

<!-- 💡 Go to the preview tab to click the links below 💡 -->

### Please confirm your submission meets all the criteria

<!-- 💡 Please replace each `[ ]` with `[X]` when the step is complete 💡 -->

- [X] Please describe the application briefly. 

This application is a third-party TUI client for [Netease Cloud Music](https://en.wikipedia.org/wiki/NetEase_Cloud_Music).
- [X] The domain used for the application ID is [controlled by the application developer(s)][appid-domain] and the [application id guidelines][appid] are followed. The author has not yet created the corresponding `.io` web page, but they should have control over it. 
- [X] I have read and followed all the [Submission requirements][reqs] and the [Submission guide][reqs2].
- [X] I have [built][build] and tested the submission locally. The application logs a warning in the terminal, but basic functionalities (music playback and login) are functional, so I'm ignoring the warning for now. 
- [X] I am an author/developer/upstream contributor to the project. If not, I contacted upstream developers about this submission. **Link: go-musicfox/go-musicfox#329** (conversation in Chinese; the developer is aware that an attempt is being made to package the program with flatpak). 

<!-- ⚠️⚠️ DO NOT modify anything below this line ⚠️⚠️  -->

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[reqs2]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

hfiguiere commented:
> The author has not yet created the corresponding .io web page, but they should have control over it.

Surely they won't have control of `io.github`. So no the app-id isn't valid.

hfiguiere commented:
btw flatpak is really not the best fit for apps that don't present a GUI

hfiguiere commented on io.github.go-musicfox.appdata.xml:
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="console-application">
+  <id>io.github.go-musicfox</id>
```
the app-id is invalid.

https://docs.flathub.org/docs/for-app-authors/requirements#application-id

lihaohong6 replied:
I just went to https://pages.github.com/ and confirmed that GitHub users can create pages with the domain `https://username.github.io.`, so this should be a valid app id because `go-musicfox` is able to create a page at url `go-musicfox.github.io` through GitHub pages anytime.

lihaohong6 commented:
> > The author has not yet created the corresponding .io web page, but they should have control over it.
> 
> Surely they won't have control of `io.github`. So no the app-id isn't valid.

I thought GitHub users have control over `<username>.github.io` in that other users are not allowed to use this subdomain, so the user `go-musicfox` should have control over `go-musicfox.github.io`. I dealt GitHub Pages a long time ago, so I might have misremembered how it works. I originally used `io.github.go-musicfox.go-musicfox` as the application id, but the repetition in the name is a bit cumbersome.

lihaohong6 commented:
If the current app id doesn't work, I can change it to `io.github.go-musicfox.go-musicfox` to conform with the examples on https://docs.flathub.org/docs/for-app-authors/requirements/#application-id, but I don't see any reason to use the more verbose version unless there's still something wrong with the current app id?

bbhtt commented:
That one is also wrong. I suggest reading the entire guideline carefully, your questions are answered there.

lihaohong6 commented:
> > If the current app id doesn't work, I can change it to io.github.go-musicfox.go-musicfox
> 
> That one is also wrong. I suggest reading the entire guideline carefully, your questions are answered there.

Thanks! I think I know what's wrong now. I added a fourth component to the app id and converted dashes to underscores. Looking through the document, I believe the app id is fine now?

bbhtt commented on io.github.go_musicfox.go_musicfox.appdata.xml:
```diff
@@ -0,0 +1,48 @@
+  <launchable type="desktop-id">io.github.go_musicfox.go_musicfox.desktop</launchable>
+  <screenshots>
+    <screenshot type="default">
+      <image>https://raw.githubusercontent.com/go-musicfox/go-musicfox/master/previews/main.png</image>
```
this should be tied to a commit not `master`.

bbhtt commented on io.github.go_musicfox.go_musicfox.yml:
```diff
@@ -0,0 +1,35 @@
+finish-args:
+    - --share=network
+    - --socket=pulseaudio
+    - --talk-name=org.freedesktop.Notifications
```
what do you use for notifications? libnotify?

lihaohong6 replied:
Yes. I'm not the author, but after tracing through the function calls, notifications are sent by executing `notify-send` using a fork of [notificator](https://github.com/0xAX/notificator).

bbhtt replied:
Shelling out to a cli tool to send notifications is weird when cross-platform APIs exist for the last 10-15 years.

Anyway, notify-send uses the portal but I do not know about kdialog.

bbhtt replied:
and multiple go bindings for libnotify exists.

bbhtt replied:
kdialogue doesn't use the portal, but anyway, if it used the APIs directly it wouldn't have been needed.

lihaohong6 replied:
The part about `kdialog` in the README file seems to be false advertising. The notification library only mentions it once in the README and doesn't call `kdialog` anywhere in the code. I agree that it is not ideal to rely on sketchy, unmaintained libraries, but migrating to some other notification library upstream would depend on how the maintainer thinks about it.

bbhtt commented on io.github.go_musicfox.go_musicfox.yml:
```diff
@@ -0,0 +1,35 @@
+        url: "https://github.com/go-musicfox/go-musicfox.git"
+        tag: v4.5.7
+        commit: 1629dd8dbaa2f433bd9011db58ba3c639a15fd0c
+      - type: file
+        path: app-icon.png
+      - type: file
+        path: io.github.go_musicfox.go_musicfox.desktop
```
these should be upstream see the note in https://docs.flathub.org/docs/for-app-authors/requirements#required-metadata

lihaohong6 replied:
The upstream icon (1024x1024 png file) is larger than flatpak's limit (512x512), while the upstream dot desktop file works but has a few minor issues. I can try work with the maintainer to (1) include a smaller version of the logo upstream and (2) figure out changes/workarounds needed for the desktop file, thought these changes will take some time to be included in the the next release, depending on what the maintainer thinks. Does that sound like a reasonable solution? I also thought about shrinking the logo with `ImageMagick` in the build process, but that seems unnecessarily complicated.

hfiguiere replied:
it's not flatpak limit. It's the standard. we just follow standards here. Just delete it in the post-install.

hfiguiere replied:
oh you mean there is only that?

bbhtt replied:
`ffmpeg -i inout.png -vf scale=${size}:${size} output.png`

lihaohong6 replied:
Thank you! Looks like the icon is all good now. I will try to resolve issues with the desktop file next.

lihaohong6 replied:
There is a minor problem with the icon file name in the upstream dot desktop file, so I used `sed` to manually set the icon. That was the only workaround required to switch to the upstream version. All issues mentioned here should be resolved now.

hfiguiere replied:
don't use sed. Use `desktop-file-edit`

lihaohong6 replied:
Thanks! I just fixed it and pushed the changes.

bbhtt commented:
> Thanks! I think I know what's wrong now. I added a fourth component to the app id and converted dashes to underscores. Looking through the document, I believe the app id is fine now?

Yes seems fine now.

hfiguiere commented:
bot, build io.github.go_musicfox.go_musicfox

bbhtt commented on io.github.go_musicfox.go-musicfox.appdata.xml:
was this submitted to https://github.com/go-musicfox/go-musicfox?

lihaohong6 replied:
It was not. I suppose I should make a PR to put it in the [deploy directory](https://github.com/go-musicfox/go-musicfox/tree/master/deploy) along with the dot desktop file? The flatpak build script can reflect that change whenever upstream releases a new stable version.

bbhtt replied:
Yes

bbhtt commented:
bot, build io.github.go_musicfox.go_musicfox

bbhtt commented on io.github.go_musicfox.go-musicfox.yml:
You have to change the filenames.

hfiguiere replied:
they are good.

bbhtt replied:
ah the title was wrong then

hfiguiere commented:
bot, build io.github.go_musicfox.go-musicfox

lihaohong6 commented:
bot, build io.github.go_musicfox.go-musicfox

bbhtt commented on io.github.go_musicfox.go-musicfox.appdata.xml:
```diff
@@ -0,0 +1,48 @@
+  <recommends>
+    <control>keyboard</control>
+  </recommends>
+  
```
https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#provides -> binary

lihaohong6 commented:
bot, build io.github.go_musicfox.go-musicfox

lihaohong6 commented:
Since builds are finally succeeding, I'll create a PR at musicfox to add the metadata file.

bbhtt commented:
Is this available in CN locale only? If so please mention that in the first `<p` tag of the description.

Otherwise how do I switch to EN?

lihaohong6 commented:
Simplified Chinese is the only supported language. I added a note in the description about it. Sorry for the confusion.

bbhtt commented:
bot, build io.github.go_musicfox.go-musicfox

bbhtt commented on io.github.go_musicfox.go-musicfox.appdata.xml:
```diff
@@ -0,0 +1,41 @@
+  </recommends>
+  <provides>
+    <binary>musicfox</binary>
+  </provides>
+  
```
```suggestion
  </provides>
  
  <content_rating type="oars-1.1" />
```

bbhtt replied:
this is mandatory.

lihaohong6 commented:
bot, build io.github.go_musicfox.go-musicfox

bbhtt commented:
/merge
